### PR TITLE
ref #650: merge basket items with identical basket item data

### PR DESCRIPTION
**Pull request by rnuernberg**
| Q             | A
| ------------- | ---
| Branch        | 7.0.x
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed issues  | chameleon-system/chameleon-system#650
| License       | MIT

### CODE REVIEW

File: src/ShopBundle/objects/TShopBasket/TShopBasketCore.class.php [hunk 1]
```diff
@@ -640,6 +640,9 @@ public function updateCustomData(string $basketIdentifier, array $customData): v
             return;
         }
         $item->setCustomData($customData);
+
+        // this may result in items with the same basket ident - in that case we should merge them
+        $this->oBasketArticles->mergeIdenticalBasketItems();
```
Comment (UlrichKu) [inline src/ShopBundle/objects/TShopBasket/TShopBasketCore.class.php:645]: The method updateCustomData seems quite innocent.

Why should it result in items being added to the basket?

File: src/ShopBundle/objects/WebModules/MTShopBasketCore/MTShopBasketCoreEndpoint.class.php [hunk 2]
```diff
@@ -941,10 +939,8 @@ public function UpdateBasketItem($aRequestData = null, $bAddAmountToExistingAmou
             } else {
                 $bItemWasUpdated = $oBasket->AddItem($oArticle);
             }
-            if (null !== $oExistingItem) {
```
Comment (UlrichKu) [inline src/ShopBundle/objects/WebModules/MTShopBasketCore/MTShopBasketCoreEndpoint.class.php:944]: Isn't this opening brace now missing? Or is it just the diff display?

File: src/ShopBundle/objects/TShopBasket/TShopBasketArticleCoreList.class.php [hunk 3]
```diff
@@ -801,4 +801,31 @@ protected function PostDeleteItemHook($oDeletedItem)
...
+     */
+    public function mergeIdenticalBasketItems(): void
+    {
+        /** @var TShopBasketArticle[] $items */
+        $items = [];
+        $mergedItems = false;
+        $this->setItemPointer(0);
+        while (false !== ($item = $this->next())) {
+            $key = $item->sBasketItemKey;
+            if (isset($items[$key])) {
```
Comment (UlrichKu) [inline src/ShopBundle/objects/TShopBasket/TShopBasketArticleCoreList.class.php:816]: array_key_exists

File: src/ShopBundle/objects/TShopBasket/TShopBasketCore.class.php (hunk 1 above)
Comment (rnuernberg) [reply src/ShopBundle/objects/TShopBasket/TShopBasketCore.class.php:645]: not at all. The basketIdentKey of an item always depends on its custom data. When a users changes the custom data of an item, they also change the basketIdentKey - which then may result in having two items with the same key.

In fact, this is exactly the problem - a user has product A with some custom data foo=bar  and the same product A with some custom data foo=baz. When the user then changes foo=baz to foo=bar in the second product, the result will be product A being twice in the basket with identical basketIdentKeys. These must now be merged

File: src/ShopBundle/objects/WebModules/MTShopBasketCore/MTShopBasketCoreEndpoint.class.php (hunk 2 above)
Comment (rnuernberg) [reply src/ShopBundle/objects/WebModules/MTShopBasketCore/MTShopBasketCoreEndpoint.class.php:944]: we removed 2 opening and one closing brace - and replaced it with one opening brace. So no, resulting number of braces is ok (moved from 2 opening and 2 closing to one opening and one closing).

File: src/ShopBundle/objects/TShopBasket/TShopBasketArticleCoreList.class.php (hunk 3 above)
Comment (rnuernberg) [reply src/ShopBundle/objects/TShopBasket/TShopBasketArticleCoreList.class.php:816]: we know that `$items` only ever contains objects - so in my eyes we have no advantage in using array_key_exists. isset has the advantage of being shorter (easier to read) and being faster.

Comment (UlrichKu) [reply src/ShopBundle/objects/TShopBasket/TShopBasketArticleCoreList.class.php:816]: And is doing more than the "array key exists" that is desired here.

File: src/ShopBundle/objects/TShopBasket/TShopBasketArticleCoreList.class.php [hunk 4]
```diff
@@ -801,4 +801,31 @@ protected function PostDeleteItemHook($oDeletedItem)
             $this->aObservers[$sObserverName]->OnBasketItemDeleteEvent($oDeletedItem);
         }
     }
+
+    /**
+     * items with the same sBasketItemKey will be merged into one item.
+     */
+    public function mergeIdenticalBasketItems(): void
```
Comment (UlrichKu) [inline src/ShopBundle/objects/TShopBasket/TShopBasketArticleCoreList.class.php:808]: Does this TShopBasketArticleCoreList reflect database entities?

If so wouldn't you need to update the db as well?

File: src/ShopBundle/objects/TShopBasket/TShopBasketCore.class.php (hunk 1 above)
Comment (UlrichKu) [reply src/ShopBundle/objects/TShopBasket/TShopBasketCore.class.php:645]: Hm, I think what I find unfortunate here:

setting custom data on an entry breaks something and requires the list to be fixed (and thus the method mergeIdenticalBasketItems there is public but needs to be called here).

Comment (rnuernberg) [reply src/ShopBundle/objects/TShopBasket/TShopBasketCore.class.php:645]: ok - that makes sense. I moved alle the logic into the article list - so we no longer need the public method mergeIdenticalBasketItems

File: src/ShopBundle/objects/TShopBasket/TShopBasketArticleCoreList.class.php (hunk 3 above)
Comment (rnuernberg) [reply src/ShopBundle/objects/TShopBasket/TShopBasketArticleCoreList.class.php:829]: if, for whatever reason, $items[$key] was null, then we would want to overwrite that array entry. So in fact, it is doing exactly what we want here :)$items[$key]

File: src/ShopBundle/objects/TShopBasket/TShopBasketArticleCoreList.class.php (hunk 4 above)
Comment (rnuernberg) [reply src/ShopBundle/objects/TShopBasket/TShopBasketArticleCoreList.class.php]: nope - the basket is kept in memory/session.

**Pull request merged**